README.md
# Oflex

O(pinionated) fl(ask) ex(tensions).

Flask login and connection-pooling stuff that I reuse in all my prototyping. The one-line goal of this project is to have drop-in login, however inflexible, rather than spending 6 hours setting it up every time.

This is only compatible with my postgres / redis stack and a similar DB users schema.

Will happily add docs if anyone else needs this, post a github issue to LMK.

## Features

These things wouldn't normally belong together and are only bound together by the common themes of (1) me needing them every time (2) I haven't found a right-fit library to do any of them in flask (3) they're all difficult to integrate automatically without some schema / template / backend assumptions.

* [x] flask blueprints for create-account and login pages
* [x] and scrypt passwords
* [x] and twilio login
* [x] login-required decorator which populates `flask.g.session`
* [x] graphene / graphql middleware login-required
* [ ] HSTS middleware
* [x] postgres & redis initializer
* [x] postgres connection pool context manager
* [ ] basic RBAC
* [ ] werkzeug remote IP
* [ ] in-memory and redis rate limiting
* [ ] email verification
* [ ] invitations & preapproval lists
* [ ] oauth
* [ ] add sample schema

## What you should do to use this

* Look in `config.py` in this repo and make sure your SQL users table and environment variables conform to what this package expects.
* (You can customize the configs, but that's a waste)
* use `flask.current_app.pool` or `pool.withcon`, and `flask.current_app.redis` in your thing
* You have a base.htm template that defines styles, headers, etc

## Consider also

These other flask login libraries that I haven't evaluated at all:

* https://github.com/mattupstate/flask-security
* https://github.com/lingthio/Flask-User
* https://github.com/maxcountryman/flask-login

There are a lot of these because schema integration is difficult.


CHANGELOG.md
# Changelog

## 0.0.x series

* 0.0.5
  - `CONFIG['login_hook']` for setting extra info in session
* 0.0.4
  - ergonomics: clearer error when sqlite doesn't exist
  - `require_verification` key in config to control invite-only mode
  - /verify + /finish flows
  - remove column name indirection
  - invite-flow.dot graphviz of new flows
* 0.0.3
  - bugfix: `include_package_data`
  - feature: sqlite support (major refactor, might break postgres support, haven't tested)
  - switch from indirect session to signed userid in cookie to remove redis dependency in some cases
* 0.0.2
  - add SMS login behind a flag (defaults to on)
  - CI / test / lint
  - configurable sql params & render\_config
  - internal: use named instead of positional vars for sql
  - internal: pytest
* 0.0.1 initial release


test/test_config.py
from oflex import config

def test_render():
  # this is confirming that there aren't any missing keys in the default config
  config.render_config()

def test_query_render_dialect():
  assert config.QueryRenderer('postgres').wildcard == '%s'
  assert config.QueryRenderer('sqlite').wildcard == '?'

def test_query_render_dialect():
  render = config.QueryRenderer('postgres')
  assert render.select('tab', ('c1',), 'c2') == "select c1 from tab where c2 = %s"
  assert render.update('tab', 'c1', 'c2') == "update tab set c1 = %s where c2 = %s"
  assert render.insert('tab', ('c1', 'c2')) == "insert into tab (c1, c2) values (%s, %s)"


oflex/pool.py
import flask, contextlib, os, redis, jinja2, time, sqlite3
from .config import render_config, CONFIG, getenv

def fetch1_abort(cur, status=404):
  "helper to flask.abort if DB query is empty"
  row = cur.fetchone()
  if row is None:
    flask.abort(status)
  return row

@contextlib.contextmanager
def withcon():
  if CONFIG['db_dialect'] == 'sqlite':
    con = sqlite3.connect(getenv('automig_con'))
    try:
      yield con
      con.commit()
    finally:
      con.close()
  else:
    pool = flask.current_app.pool
    con = pool.getconn()
    try:
      yield con
      con.commit()
    finally:
      pool.putconn(con)

def init():
  render_config()
  app = flask.current_app
  if CONFIG['db_dialect'] == 'postgres':
    # todo: register postgres uuid
    import psycopg2.pool # delayed so this isn't a hard dep
    app.pool = psycopg2.pool.ThreadedConnectionPool(0, CONFIG['maxconn'], getenv('automig_con'))
  elif CONFIG['db_dialect'] == 'sqlite':
    assert os.path.exists(getenv('automig_con')), "sqlite DB path doesn't exist, did you forget to init?"
  else:
    raise ValueError('unk dialect', CONFIG['db_dialect'])
  if CONFIG['use_redis']:
    app.redis = redis.Redis(getenv('redis'))
  if CONFIG['support_sms']:
    import twilio.rest # delayed so this isn't a hard dep
    app.twilio = twilio.rest.Client(getenv('twilio_sid'), getenv('twilio_token'))
  app.jinja_loader = jinja2.ChoiceLoader([
    app.jinja_loader,
    jinja2.FileSystemLoader([os.path.join(os.path.dirname(__file__), 'templates')]),
  ])


oflex/config.py
import yaml, os, collections, flask

class QueryRenderer:
  "poor man's ORM"

  def __init__(self, dialect):
    self.dialect = dialect
    self.wildcard = '?' if self.dialect == 'sqlite' else '%s'

  def select(self, table, sel_cols, wherecol):
    return f"select {', '.join(sel_cols)} from {table} where {wherecol} = {self.wildcard}"

  def update(self, table, setcol, wherecol):
    return f"update {table} set {setcol} = {self.wildcard} where {wherecol} = {self.wildcard}"

  def insert(self, table, cols):
    return f"insert into {table} ({', '.join(cols)}) values ({', '.join([self.wildcard] * len(cols))})"

def default_send_verify(email, verification_code):
  print('verify', flask.url_for('oflex.blueprint.get_verify', email=email, verification_code=verification_code, _external=True))

RAW_CONFIG = dict(
  queries=None,
  # named fields namedtuple for select + insert returning
  query_fields=dict(
    get_user_email=collections.namedtuple('get_user_email', 'auth_method pass_hash pass_salt userid'),
    get_user_email_ver=collections.namedtuple('get_user_email_ver', 'auth_method pass_hash pass_salt userid verified'),
    get_user_sms=collections.namedtuple('get_user_sms', 'auth_method userid username'),
    get_username=collections.namedtuple('get_username', 'username'),
  ),
  session_expiry=86400 * 90,
  # maximum size of connection pool
  maxconn=4,
  db_dialect='postgres',
  use_redis=True,
  env=dict(
    # name of environment variable that holds postgres connection string
    automig_con='AUTOMIG_CON',
    redis='REDIS_HOST',
    twilio_sid='TWILIO_SID',
    twilio_token='TWILIO_TOKEN',
    twilio_from='TWILIO_FROM',
  ),
  # route to visit after successful login
  login_home='home',
  # maxlen
  max_email=400,
  support_sms=True,
  appname='oflex',
  # username comment is editable so sites can give guidance about how this will be used
  username_name='Username',
  username_comment='(public)',
  # require_verification means to have 'verified' and 'verification_code' fields for email login
  require_verification=False,
  send_verification_email=default_send_verify,
  login_hook=lambda kind, row: None,
  event_hook=lambda kind, userid: None,
)

# note: this is a dict rather than None so import refs work
CONFIG = {}

def load_config(path='oflex.yml'):
  "load config from file"
  # todo: deep merge instead of update
  RAW_CONFIG.update(yaml.load(open(path)))

def render_config():
  "render queries"
  global CONFIG
  tmp = RAW_CONFIG.copy() # this isn't a deep copy -- why am I bothering? what's RAW_CONFIG -> CONFIG about?
  assert tmp['db_dialect'] in ('postgres', 'sqlite')
  render = QueryRenderer(tmp['db_dialect'])
  if tmp['require_verification']:
    # ugh: this is going to cause problems someday
    tmp['query_fields']['get_user_email'] = tmp['query_fields']['get_user_email_ver']
  tmp['queries'] = dict(
    create_user_email=render.insert('users', ('userid', 'username', 'auth_method', 'email', 'pass_hash', 'pass_salt')),
    create_user_email_ver=render.insert('users', ('userid', 'username', 'auth_method', 'email', 'pass_hash', 'pass_salt', 'verification_code')),
    create_user_sms=render.insert('users', ('userid', 'auth_method', 'sms')),
    update_username=render.update('users', 'username', 'userid'),
    get_verify=render.select('users', ('verification_code', 'verified', 'pass_hash', 'userid'), 'email'),
    # note: get_* select queries are using namedtuple fields. so much ugly indirection here, just use an ORM
    get_user_email=render.select('users', tmp['query_fields']['get_user_email']._fields, 'email'),
    get_user_sms=render.select('users', tmp['query_fields']['get_user_sms']._fields, 'sms'),
    get_username=render.select('users', tmp['query_fields']['get_username']._fields, 'userid'),
  )
  CONFIG.update(tmp)

def getenv(name):
  "get env var with indirection"
  return os.environ[CONFIG['env'][name]]


oflex/middleware.py
import flask, json, functools
from datetime import datetime

def require_session(inner):
  "decorator to require login on a session. use beneath the @route() decorator or it won't work"
  if hasattr(inner, '__wrapped__'):
    print("warning: require_session is wrapping a function that already has been wrapped. Make sure you're using it *inside* the @route decorator", inner)
  @functools.wraps(inner)
  def outer(*args, **kwargs):
    # todo: remember redirect page
    userid = flask.session.get('userid')
    # note: 'expires' here should never matter b/c cookie expires in 31 days, but this is signed and that isn't I think so I have this
    if userid is None or 'expires' not in flask.session or flask.session['expires'] < datetime.now():
      flask.flash('Session expired, sign in again')
      return flask.redirect(flask.url_for('oflex.blueprint.get_login'))
    flask.g.session = {'userid': userid}
    return inner(*args, **kwargs)
  return outer


oflex/blueprint.py
import flask, uuid, json, scrypt, os, random, binascii, logging, urllib
from datetime import datetime, timedelta
from . import pool, middleware
from .config import CONFIG, getenv

APP = flask.Blueprint(__name__, 'oflex')

@APP.route('/login')
def get_login():
  return flask.render_template('login.htm', support_sms=CONFIG['support_sms'])

def set_session(userid):
  flask.session['userid'] = userid
  flask.session['expires'] = datetime.now() + timedelta(days=90) # used in middleware
  flask.session.permanent = True # i.e. permanent_session_lifetime = 31 days default. user can probably override this, expiration isn't secure.

def tobytes(raw):
  "helper to return bytes whether bytes or memoryview (postgres / sqlite)"
  return raw if isinstance(raw, bytes) else raw.tobytes()

@APP.route('/login/email', methods=['POST'])
def post_login_email():
  "login with email"
  form = flask.request.form
  with pool.withcon() as con:
    cur = con.cursor()
    cur.execute(CONFIG['queries']['get_user_email'], (form['email'],))
    raw_row = cur.fetchone()
    if raw_row is None:
      flask.flash('Bad login info!')
      return flask.redirect(flask.url_for('oflex.blueprint.get_login'))
  row = CONFIG['query_fields']['get_user_email'](*raw_row)
  assert row.auth_method == 'email'
  if tobytes(row.pass_hash) != scrypt.hash(form['password'].encode(), tobytes(row.pass_salt)):
    flask.flash('Bad login info!')
    return flask.redirect(flask.url_for('oflex.blueprint.get_login'))
  if CONFIG['require_verification'] and not row.verified:
    flask.flash("That password is correct but your email hasn't been verified yet!")
    return flask.redirect(flask.url_for('oflex.blueprint.get_wait'))
  set_session(row.userid)
  CONFIG['login_hook']('email', row)
  CONFIG['event_hook']('login_email', row.userid)
  return flask.redirect(flask.url_for(CONFIG['login_home']))

@APP.route('/wait')
def get_wait():
  return flask.render_template('wait_verify.htm')

@APP.route('/verify')
def get_verify():
  email = flask.request.args['email']
  if os.environ.get('DOUBLE_DECODE'):
    # yup, LB + lambda = double encoding. https://github.com/dougmoscrop/serverless-http/issues/74
    email = urllib.parse.unquote(email)
  verification_code = flask.request.args['verification_code']
  if not email or not verification_code:
    flask.abort(flask.Response("Missing query params in link", status=400))
  with pool.withcon() as con:
    cur = con.cursor()
    cur.execute('begin')
    cur.execute(CONFIG['queries']['get_verify'], (email,))
    row = cur.fetchone()
    unk_details = flask.Response("Unknown verification details", status=404)
    if not row:
      logging.debug('unk email %s', email)
      flask.abort(unk_details) # possible attack
    actual_code, verified, pass_hash, userid = row
    if verified is not None and actual_code == verification_code:
      flask.abort(flask.Response("Already verified!", status=400))
    if verified is not None or actual_code != verification_code:
      logging.debug('already verified (%s) or bad code (%s)', verified is not None, actual_code != verification_code)
      flask.abort(unk_details) # possible attack
    assert verified is None and actual_code == verification_code
    cur.execute('update users set verified = now() where email = %s', (email,))
    cur.execute('commit')
  CONFIG['event_hook']('verify', userid)
  if pass_hash is None:
    # note: this is the pre-pass case; oflex can't generate this case, it requires an invite flow initiated by the app
    # todo: add an invite() function somewhere to do this
    return flask.redirect(flask.url_for('oflex.blueprint.get_finish', email=email, verification_code=verification_code))
  else:
    # note: has-pass case
    flask.flash('Verification successful! Now log in')
    return flask.redirect(flask.url_for('oflex.blueprint.get_login'))

@APP.route('/finish')
def get_finish():
  assert CONFIG['require_verification']
  return flask.render_template('finish.htm', email=flask.request.args['email'], verification_code=flask.request.args['verification_code'])

@APP.route('/finish', methods=['POST'])
def post_finish():
  form = flask.request.form
  pass_salt = os.urandom(64)
  pass_hash = scrypt.hash(form['password'].encode(), pass_salt)
  with pool.withcon() as con:
    cur = con.cursor()
    cur.execute('select userid, verification_code from users where email = %s', (form['email'],))
    userid, actual_code = cur.fetchone()
    if form['verification_code'] != actual_code:
      flask.abort(flask.Response("Bad verification code; make sure you followed the invite link correctly. Ask your admin to send you another invitation maybe.", status=400))
    cur.execute('update users set pass_salt=%s, pass_hash=%s where userid=%s', (pass_salt, pass_hash, userid))
  flask.flash("Password set successfully! Now log in")
  return flask.redirect(flask.url_for('oflex.blueprint.get_login'))

@APP.route('/join')
def get_join():
  return flask.render_template('join.htm', support_sms=CONFIG['support_sms'], username_comment=CONFIG['username_comment'], username_name=CONFIG['username_name'])

@APP.route('/join/email', methods=['POST'])
def post_join_email():
  "create account with email / password"
  form = flask.request.form
  if not form['password']:
    flask.abort(flask.Response("Blank password not allowed", status=400))
  assert '@' in form['email'] and len(form['email']) < CONFIG['max_email']
  pass_salt = os.urandom(64)
  pass_hash = scrypt.hash(form['password'].encode(), pass_salt)
  userid = str(uuid.uuid4())
  with pool.withcon() as con:
    cur = con.cursor()
    # todo: rate limiting
    # todo: clearer error for duplicate email
    if CONFIG['require_verification']:
      verification_code = binascii.hexlify(os.urandom(10)).decode()
      cur.execute(
        CONFIG['queries']['create_user_email_ver'],
        (userid, form['username'], 'email', form['email'], pass_hash, pass_salt, verification_code)
      )
      CONFIG['send_verification_email'](form['email'], verification_code)
    else:
      cur.execute(
        CONFIG['queries']['create_user_email'],
        (userid, form['username'], 'email', form['email'], pass_hash, pass_salt)
      )
      # note: only set_session in this !require_verification branch
      set_session(userid)
  CONFIG['event_hook']('join_email', userid)
  if CONFIG['require_verification']:
    return flask.redirect(flask.url_for('oflex.blueprint.get_wait'))
  else:
    return flask.redirect(flask.url_for(CONFIG['login_home']))

@APP.route('/logout', methods=['POST'])
def post_logout():
  flask.session.clear()
  return flask.redirect(flask.url_for('oflex.blueprint.get_login'))

@APP.route('/login/sms')
def get_login_sms():
  assert CONFIG['support_sms']
  return flask.render_template('sms.htm')

def send_sms_code(number):
  "aborts if there's a problem. doesn't hit twilio in FLASK_DEBUG mode"
  import phonenumbers # delayed for soft dep
  try: number = phonenumbers.parse(number, 'US')
  except Exception as err:
    flask.abort(flask.Response(f"Problem parsing phone number: {str(err)}", status=400))
  if not phonenumbers.is_valid_number(number):
    flask.abort(flask.Response("Invalid number", status=400))
  if number.country_code != 1:
    flask.abort(flask.Response("Non-US number -- we only support US phone numbers for now. Contact support to talk about other use-cases.", status=400))
  flask.session['sms'] = normal = phonenumbers.format_number(number, phonenumbers.PhoneNumberFormat.E164)
  secret_code = '%06d' % int(random.random() * 1e6)
  if os.environ.get('FLASK_DEBUG') == '1':
    print('fake local twilio:', secret_code)
  else:
    # todo: rate limit
    flask.current_app.twilio.messages.create(
      body=f"Your {CONFIG['appname']} verification code is: {secret_code}.\n\nReply STOP to unsubscribe.",
      from_=getenv('twilio_from'),
      to=normal,
    )
  flask.current_app.redis.setex(f'sms-{normal}', 60 * 30, secret_code)

@APP.route('/login/sms', methods=['POST'])
def post_login_sms():
  assert CONFIG['support_sms']
  send_sms_code(flask.request.form['sms'])
  return flask.redirect(flask.url_for('oflex.blueprint.get_confirm'))

@APP.route('/login/sms/confirm')
def get_confirm():
  assert CONFIG['support_sms']
  return flask.render_template('confirm.htm')

@APP.route('/login/sms/confirm', methods=['POST'])
def post_confirm():
  assert CONFIG['support_sms']
  normal = flask.session['sms']
  received_code = flask.request.form['confirm']
  confirm_key = f'sms-{normal}'
  correct_code = flask.current_app.redis.get(confirm_key)
  if received_code.encode() != correct_code:
    flask.abort(flask.Response('Bad code. Try again or request another code', status=401))
  dest = flask.url_for(CONFIG['login_home'])
  with pool.withcon() as con:
    cur = con.cursor()
    # todo: rate limiting
    cur.execute(CONFIG['queries']['get_user_sms'], (normal,))
    row = cur.fetchone()
    row = row and CONFIG['query_fields']['get_user_sms'](*row)
    userid = row and row.userid
    if not row:
      userid = str(uuid.uuid4())
      cur.execute(CONFIG['queries']['create_user_sms'], (userid, 'sms', normal))
      row = cur.fetchone()
      dest = flask.url_for('oflex.blueprint.username')
    elif not row.username:
      dest = flask.url_for('oflex.blueprint.username')
    set_session(userid)
  flask.current_app.redis.delete(confirm_key)
  return flask.redirect(dest)

@APP.route('/login/username')
def username():
  assert CONFIG['support_sms']
  return flask.render_template('username.htm', username_comment=CONFIG['username_comment'])

@APP.route('/login/username', methods=['POST'])
@middleware.require_session
def post_username():
  assert CONFIG['support_sms']
  username = flask.request.form['username']
  if len(username) < 3:
    flask.abort(flask.Response('Username too short -- 3 characters minimum', status=400))
  if len(username) > 64:
    flask.abort(flask.Response('Username too long -- 64 characters max', status=400))
  with pool.withcon() as con:
    cur = con.cursor()
    cur.execute(CONFIG['queries']['get_username'], (flask.g.session['userid'],))
    row = CONFIG['query_fields']['get_username'](*cur.fetchone())
    if row.username:
      flask.abort(flask.Response("Can't set a username! You already have one", status=400))
    cur.execute(CONFIG['queries']['update_username'], (username, flask.g.session['userid']))
  return flask.redirect(flask.url_for(CONFIG['login_home']))
